File: Python/Picas&Fijas.py
#!/bin/python3
         
#                         #                        
#                       @@@@#                      
#                     %@@@@@@@                     
#                    @@@@@@@@@@@                   
#                  @@@@@@@@@@@@@@#                
#                @@@@@@@@@@@@@@@@@@.               
#              &@@@@@@@@@@@@@@@@@@@@@              
#            @@@@@@@@@@@@@@@@@@@@@@@@@@            
#           @@/       \@@@@@@@/       \@@          ~ ch4rum
#         @@@@    __   \@@@@@@    _____|@@@        ~ https://github.com/ch4rum/ProgrammingExer
#        @@@@@   |@@|  |@@@@@@   |@@@@@@@@@@       ~ python3
#      %@@@@@@   ---   |@@@@@@      \@@@@@@@@      
#      @@@@@@@    ____/@@@@@@@    __|@@@@@@@@@     
#     .@@@@@@@   |@@@@@@@@@@@@   |@@@@@@@@@@@@     
#      @@@@@@@   |@@@@@ @ @@@@   |@@@@@@@@@@@@     
#       @@@@@@\__|@@@@  @ @@@@\__|@@@@@@@@@@.      
#         @@@@@@@@@@@@@@ @@@@@@@@@@@@@@@@@@        
#                    @@@@@@@@@@@                   
#                   @@@@@@@@@@@@@                  
#                  @@@@@@@@@@@@@@@                 

import random, signal, sys, os

def handler(sig, frame):
    print(f"\n\n[!] Saliendo ...\n");
    sys.exit(1);

# Ctrl +C
signal.signal(signal.SIGINT, handler);

def generateKey():
    # Genera la contraseña de 4 digitos aleatoria sin repetir decimales
    key = random.sample(range(0,10), 4);
    return key;

def get_picOrFij(key,attempt):
    # Funcion que verifica si es fija o pica; compara en el rango de 4 si esta en la
    # misma posicion es fija o si esta en el numero es pica, retorna picas &  fijas
    picas, fijas = 0, 0;
    for values in range(4):
        if attempt[values] == key[values]: fijas += 1;
        elif attempt[values] in key: picas += 1;
    return picas, fijas;

def showMSJ(count):
    # Muestra un mensaje segun los intentos en que haya adivinado el numero
    if count < 2:
        print(f"* Exelente, eres un maestro estas fuera del alcance de los demas\n");
    elif count < 4:
        print(f"* Muy bueno, puedes ser un gran competidor\n");
    elif count < 8:
        print(f"* Bien, estas progresando debes buscar tus limites\n");
    elif count < 10:
        print(f"* Regular, Aun es largo el camino por recorrer\n");
    else:
        print(f"* Mal, este juego no es para ti\n");

def export(key,count,date_attempt):
    # NewFuncion para exportar el historial de una partida, primero verifica
    # el sistema operativo y si una ruta existe sino la crea, luego abre o crea
    # un nuevo archivo para poner los datos de la partida y guardarlos.
    if os.name == "nt":
        desktop = f"C:\\Users\\{os.getlogin()}\\Desktop\\";
        if not os.path.exists(desktop):
            os.makedirs(desktop);
    elif os.name == "posix":
        desktop = f"/home/{os.getlogin()}/Desktop/";
        if not os.path.exists(desktop):
            os.makedirs(desktop);

    export = open(desktop + "Picas&Fijas.txt", "a");
    export.write(f"\tNueva partida :)\n- Intento: [{count}]\n- Key: {key}\n- Datos intentos: ");
    for values in date_attempt:
        export.write(f"{values}\n");
    export.write("\n\n");
    export.close();
    print(f"Archivo exportado exitosamente! --> {desktop}Picas&Fijas.txt\n");

if __name__ == "__main__":
    print(f"\n\t♠ Picas & Fijas  ♠\n")
    key = generateKey();
    #print(key);
    count = 0; date_attempt = [];
    while count < 12:
        attempt = input("* Ingrese un numero de 4 cifras sin repetir decimales :> ");
        if len(attempt) != 4 or not attempt.isdigit():
            #print(f"Ingrese un numero de 4 cifras sin repetir decimales!");
            continue;
        attempt = [int(values) for values in attempt];
        #for values in attempt:
        #    attempt.append(int(values));
        if len(set(attempt)) < 4:
            #print(f"Ingresa un numero de 4 cifras sin repetir decimales !");
            continue;
        picas , fijas = get_picOrFij(key, attempt);
        count += 1;
        date_attempt.append(attempt);
        print(f"\n- Intento: [{count}]\n- Picas: [{picas}]\n- Fijas: [{fijas}]\n");
        if fijas == 4:
            showMSJ(count);
            break;
        if count == 12:
            showMSJ(count);
            #key_str = "";
            #for values in key:
            #    key_str += str(values);
            print(f"\nLa clave era : {''.join(str(values) for values in key)}\n");
    export(key, count, date_attempt);

File: C++/Picas&Fijas.cpp
/*         
                         #                        
                       @@@@#                      
                     %@@@@@@@                     
                    @@@@@@@@@@@                   
                  @@@@@@@@@@@@@@#                
                @@@@@@@@@@@@@@@@@@.               
              &@@@@@@@@@@@@@@@@@@@@@              
            @@@@@@@@@@@@@@@@@@@@@@@@@@            
           @@/       \@@@@@@@/       \@@          ~ ch4rum
         @@@@    __   \@@@@@@    _____|@@@        ~ https://github.com/ch4rum/ProgrammingExer
        @@@@@   |@@|  |@@@@@@   |@@@@@@@@@@       ~ python3
      %@@@@@@   ---   |@@@@@@      \@@@@@@@@      
      @@@@@@@    ____/@@@@@@@    __|@@@@@@@@@     
     .@@@@@@@   |@@@@@@@@@@@@   |@@@@@@@@@@@@     
      @@@@@@@   |@@@@@ @ @@@@   |@@@@@@@@@@@@     
       @@@@@@\__|@@@@  @ @@@@\__|@@@@@@@@@@.      
         @@@@@@@@@@@@@@ @@@@@@@@@@@@@@@@@@        
                    @@@@@@@@@@@                   
                   @@@@@@@@@@@@@                  
                  @@@@@@@@@@@@@@@                 
*/

#include<iostream>
#include<stdlib.h>
#include<time.h>
#include<string>
#include<vector>
#include<set>

using namespace std;

pair<int, int> get_picOrFij(const string& key,const string& attempt){
    /* Funcion que verifica cuales son picas y cuales son fijas en el 
    intento del usuario, retorna las 2 variables de picas ++, fijas ++,
    recordando que fijas es que esten en la misma pocision de la llave y 
    picas en diferente posicion.*/
    int picas = 0, fijas = 0;
    for (int _=0;_<4;_++){
        for (int l=0;l<4;l++){
            if (key[_] == attempt[l]){
                picas++;
            }
        }
    }
    for (int _=0;_<4;_++){
        if (key[_] == attempt[_]){
            fijas++;
        }
    }
    return make_pair(picas,fijas);
}

string generateKey(){
    /* Funcion que genera el numero aleatorio y lo retorna, cumpliendo
    con que no se repita el numero de la llave*/
    srand(time(NULL));
    string(key);

    for(int _=0;_<4;_++){
        int ramdon = rand()%10;
        bool duplicate = false;
        for (int j=0;j<_;j++){
            if (key[j] == static_cast<char>(ramdon + '0')) {
                duplicate = true;
                break;
            }
        }
        if (!duplicate){
            key += static_cast<char>(ramdon + '0');
        } else{
            _--;
        }
    }
    return key;
}

void showMSJ(int count){
    /* Funcion que solo muestra un mensaje segun los intentos del usuario*/
    if (count<2){
        cout<<"* Excelente,eres un maestro estas fuera del alcance de los demas"<<endl;
    } else if (count<4){
        cout<<"* Muy bueno, puedes ser un gran competidor"<<endl;
    } else if (count<8){
        cout<<"* Bien, estas progresando debes buscar tus limites"<<endl;
    } else if (count<10){
        cout<<"* Regular, Aun es largo el camino por recorrer"<<endl;
    } else {
        cout<<"* Mal, este juego no es para ti"<<endl;
    }
}

int main (){
    /* Inicio del juego */
    string attempt, key;
    int count = 0, picas = 0, fijas = 0;
    key = generateKey();
    cout<<"\n\t♠ Picas & Fijas  ♠\n"<<endl;
    //cout<<key<<endl; // Descomenta si quieres ver la llave
    while (count < 12){
        cout<<"* Ingrese un numero de 4 cifras sin repetir decimales :> ";cin>>attempt;
        if ((attempt.length() != 4) || (!isdigit(attempt[0])) || (!isdigit(attempt[1]))|| (!isdigit(attempt[2]))|| (!isdigit(attempt[3]))){
            continue;
        }
        vector<int> attemptVec;
        for (char digit : attempt){
            attemptVec.push_back(digit - '0');
        }
        if (set<int>(attemptVec.begin(), attemptVec.end()).size() < 4){
            continue;
        }
        count++;

        pair<int, int> result = get_picOrFij(key, attempt);
        picas = result.first;
        fijas = result.second;

        cout<<"\n- Intento: ["<<count<<"]\n- Picas: ["<<picas<<"]\n- Fijas: ["<<fijas<<"]\n"<<endl;
        if (fijas == 4){
            showMSJ(count);break;
        }
        if (count == 12){
            showMSJ(count);
            cout<<"\nLa clave era: "<<key<<endl;
        }
    }
    return 0;
}